### PR TITLE
build(client): Omit "internal" APIs from type tests

### PR DESCRIPTION
**Pull request by tylerbutler**
_This was pulled out of #22392 so that PR has more discussion._

This change updates the typetest entrypoints for packages to be the lowest-level/broadest available entrypoint other than internal. The intent of this change is to stop type testing internal APIs, but the design is deliberately entrypoint-focused so any entrypoint can be tested. Indeed, as a follow-up to this change, package owners can adjust the entrypoint tested as their requirements change.

Towards the goal of no longer testing internal APIs, The typetests have been regenerated with the updated entrypoints.

Rough logic:

- If the package has a ./legacy export, entrypoint = legacy.
- Else if the package has an ./alpha export, entrypoint = alpha.
- Else if the package has a ./beta export, entrypoint = beta.
- Else if the package has a . export, entrypoint = public.

Because of the current API hierarchy, legacy is _usually_ a superset of alpha/beta/public. This may not be true in the future as we refine the legacy API and move them between levels; package configurations will need to be adjusted for such packages.

### public

packages/common/client-utils
packages/common/driver-definitions
packages/dds/tree
packages/framework/fluid-static

### beta

packages/service-clients/tinylicious-client

### alpha

packages/tools/devtools/devtools-core
packages/tools/devtools/devtools

### legacy

azure/packages/azure-service-utils
experimental/dds/attributable-map
packages/common/container-definitions
packages/common/core-interfaces
packages/common/core-utils
packages/dds/counter
packages/dds/map
packages/dds/matrix
packages/dds/merge-tree
packages/dds/ordered-collection
packages/dds/register-collection
packages/dds/sequence
packages/dds/shared-object-base
packages/dds/shared-summary-block
packages/dds/task-manager
packages/drivers/debugger
packages/drivers/driver-web-cache
packages/drivers/local-driver
packages/drivers/odsp-driver-definitions
packages/drivers/odsp-driver
packages/drivers/routerlicious-driver
packages/framework/agent-scheduler
packages/framework/aqueduct
packages/framework/request-handler
packages/framework/synthesize
packages/loader/container-loader
packages/loader/driver-utils
packages/runtime/container-runtime-definitions
packages/runtime/container-runtime
packages/runtime/datastore-definitions
packages/runtime/datastore
packages/runtime/id-compressor
packages/runtime/runtime-definitions
packages/runtime/runtime-utils
packages/runtime/test-runtime-utils
packages/service-clients/azure-client
packages/test/test-utils
packages/tools/fluid-runner
packages/utils/odsp-doclib-utils
packages/utils/telemetry-utils

### CODE REVIEW

**Comment by tylerbutler** [comment]
@markfields FYI since you've inquired about this.

**Comment by alexvy86** [comment]
> highest available entrypoint other than internal

Just calling out that I think of them the other way around, with internal being the "lowest", then legacy, alpha, beta, public (I'm aware that legacy is orthogonal, ignoring that for the moment and for the logic used inthe PR). Maybe broadest->narrowest is less ambiguous? For further discussions about entrypoints.

Besides that, a note that I've seen @CraigMacomber is working on exposing `@alpha` APIs for tree so that one might need to switch from public to alpha.

**Comment by anthony-murphy** [comment]
I have some concerns about the approach here in that our tagging structure isn't purely hierarchical as i understand it, specifically legacy does not include alpha, and i'm not sure about beta. How is this handled? I also worry about the need to manually manage these levels, and how that could easily lead to misses. Did we consider a simpler approach of just excluding api's tagged internal?

**Comment by tylerbutler** [comment]
> Just calling out that I think of them the other way around, with internal being the "lowest", then legacy, alpha, beta, public (I'm aware that legacy is orthogonal, ignoring that for the moment and for the logic used inthe PR). Maybe broadest->narrowest is less ambiguous? For further discussions about entrypoints.

That's fair - "highest level of support - highest requirements for change" - that's one wayto think about it. Another way is "highest=most APIs exposed", which is how I think about it. Curious what others think by default. My guess is yours is more typical.

> I have some concerns about the approach here in that our tagging structure isn't purely hierarchical as i understand it, specifically legacy does not include alpha, and i'm not sure about beta. How is this handled?

This is a benefit to the design - the types tested are whatever types are exposed via the configured entrypoint. That means the type tests don't need to know anything about the relationship between the entrypoints and "API levels". Type tests just deal with entrypoints.

A limitation of the current implementation is that every invocation of generate:typetests generates one file, and corresponds to one API level. This was deliberate for simplicity; if we want to test multiple entrypoints, we generate tests for each entrypoint. That said, it is relatively easy to update the config and CLI to take an array of entrypoints instead of a single one, and generate tests for all types from all entrypoints. If/when that is needed it's on the table.

> I also worry about the need to manually manage these levels, and how that could easily lead to misses. Did we consider a simpler approach of just excluding api's tagged internal?

We considered a couple of alternatives, but "excluding internal" is not straightforward because our packages by design do not all expose the same entrypoints. Some have alpha, some have beta, etc. And that changes over time as you noted. One design we considered was to forgo the config and always pick the "lowest" level entrypoint that a package has except for internal. But that means that all packages with alpha APIs, for example, will have them type tested. Today, folks have asked to not test internal APIs - tomorrow, an experimental package will want to omit alpha APIs, etc. The current design provides that flexibility without baking that logic into the code. But as you noted, a drawback is that it has to be "managed" in some way.

Another advantage to the entrypoint-based design is that we can easily type test arbitrary entrypoints. As we refine our APIs, I envision us starting to expose different entrypoints to help organize our APIs. FOr example, in fluid-0framework we re-export all of tree, all of map, etc. It's a big jumble of top-level APIs. But if we exposed the tree apis at fluid-framework/tree, for example, that might be easier for customer to understand and use. Just one example, but the overall point is that entrypoints represent the way customers see our APIs, so typing type test generation to them seems correct.

Thinking about the future - I am working to get to a point where we can include build-tools in the same workspace as the client packages. Once we can do that, I'm much more amenable to code-only solutions because we can change the code quickly without releasing new build-tools versions. As it is, though, making things configurable gives us more wiggle room to handle changing/unclear requirements without constantly releasing and integrating build-tools.

**Review by anthony-murphy — APPROVED** [review]

**Review by CraigMacomber — APPROVED** [review]
Approving changes to tree. Tree is just migrating from the old way to declare its testing public to the new way, so no issue there.

File: packages/tools/devtools/devtools-core/src/test/types/validateDevtoolsCorePrevious.generated.ts [hunk 1]
```diff
@@ -15,60 +15,6 @@ import type * as current from "../../index.js";
 
 declare type MakeUnusedImportErrorsGoAway<T> = TypeOnly<T> | MinimalType<T> | FullType<T> | typeof old | typeof current | requireAssignableTo<true, true>;
 
-/*
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-core/src/test/types/validateDevtoolsCorePrevious.generated.ts:18]: Woooo!!! ❤️🥳🥂🙌🍾🎇🎆

**Review by Josmithr — APPROVED** [review]
Devtools changes look good. No... great! No... whatever is above great!!! Thanks so much for enabling this!!!!

**Pull request merged**